agent/containermetadata/metadata_manager.go

```diff
@@ -0,0 +1,151 @@
+	inspectContainerTimeout     = 30 * time.Second
+)
+
+type MetadataStatus string
+
+const (
+	Metadata_NOT_READY = "NOT_READY"
+	Metadata_READY     = "READY"
+)
+
+// MetadataManager is an interface that allows us to abstract away the metadata
+// operations
+type MetadataManager interface {
+	SetContainerInstanceARN(string)
+	CreateMetadata(*docker.Config, *docker.HostConfig, *api.Task, *api.Container) error
+	UpdateMetadata(string, *api.Task, *api.Container) error
+	CleanTaskMetadata(*api.Task) error
+}
+
+// metadataManager implements the MetadataManager interface
+type metadataManager struct {
+	// client is the Docker API Client that the metadata manager uses. It defaults
+	// to 1.21 on Linux and 1.24 on Windows
+	client dockerMetadataClient
+	// cluster is the cluster where this agent is run
+	cluster string
+	// dataDir is the directory where the metadata is being written. For Linux
+	// this is a container directory
+	dataDir string
+	// dataDirOnHost is the directory from which dataDir is mounted for Linux
+	// version of the agent
+	dataDirOnHost string
+	// containerInstanceARN is the Container Instance ARN registered for this agent
+	containerInstanceARN string
+}
+
+// NewMetadataManager creates a metadataManager for a given DockerTaskEngine settings.
+func NewMetadataManager(client dockerMetadataClient, cfg *config.Config) MetadataManager {
+	return &metadataManager{
+		client:        client,
+		cluster:       cfg.Cluster,
+		dataDir:       cfg.DataDir,
+		dataDirOnHost: cfg.DataDirOnHost,
+	}
+}
+
+// SetContainerInstanceARN sets the metadataManager's ContainerInstanceArn which is not available
+// at its creation as this information is not present immediately at the agent's startup
+func (manager *metadataManager) SetContainerInstanceARN(containerInstanceARN string) {
+	manager.containerInstanceARN = containerInstanceARN
+}
+
+// Createmetadata creates the metadata file and adds the metadata directory to
+// the container's mounted host volumes
+// Pointer hostConfig is modified directly so there is risk of concurrency errors.
+func (manager *metadataManager) CreateMetadata(config *docker.Config, hostConfig *docker.HostConfig, task *api.Task, container *api.Container) error {
+	// Do not create metadata file for internal containers
+	if container.IsInternal {
+		return nil
+	}
+
+	// Create task and container directories if they do not yet exist
+	metadataDirectoryPath, err := getMetadataFilePath(task, container, manager.dataDir)
+	// Stop metadata creation if path is malformed for any reason
+	if err != nil {
+		return fmt.Errorf("container metadata create: %v", err)
+	}
+
+	err = os.MkdirAll(metadataDirectoryPath, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("creating metadata directory for task %s: %v", task, err)
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := manager.parseMetadataAtContainerCreate(task.Arn, container.Name)
+	data, err := json.MarshalIndent(metadata, "", "\t")
```

Comment:
While I have handled this error, I'm not sure if it is possible for an error to occur here and whether it is worthwhile to test for it. Since all my fields in the Metadata struct are strings, there should never be an error in the Marshaling. https://stackoverflow.com/questions/33903552/what-input-will-cause-golangs-json-marshal-to-return-an-error